crates/libp2p-networking/src/network/behaviours/dht/mod.rs

```diff
@@ -470,7 +476,39 @@ impl DHTBehaviour {
             KademliaEvent::UnroutablePeer { peer } => {
                 debug!("Found unroutable peer {:?}", peer);
             }
-            KademliaEvent::InboundRequest { request: _r } => {}
+            KademliaEvent::InboundRequest { request } => {
+                if let InboundRequest::PutRecord {
+                    source: _s,
+                    connection: _c,
+                    record: Some(record),
+                } = request
+                {
+                    // If the record is expired, ignore it
+                    if record.is_expired(Instant::now()) {
+                        warn!("Received an expired record");
+                        return None;
+                    }
+
+                    // Convert the record to the correct type
+                    if let Ok(record_value) = RecordValue::<K>::try_from(record.clone()) {
+                        // Convert the key to the correct type
+                        let Ok(record_key) = RecordKey::try_from_bytes(&record.key.to_vec()) else {
+                            warn!("Failed to deserialize record key");
+                            return None;
+                        };
+
+                        // If the record is signed by the correct key,
+                        if record_value.validate(&record_key) {
+                            // Store the record
+                            if let Err(err) = store.put(record.clone()) {
+                                warn!("Failed to store record: {:?}", err);
+                            }
+                        } else {
+                            warn!("Failed to validate record");
+                        }
```

Comment:
I don't think this is going to do what we want.  I'm fairly certain that the DHT is storing all records automatically when we get a put request.  I think this event is emitted to do something extra on the inbound request but at this point I think the record is likely already stored.

Reply:
Maybe an easy fix for this is to remove from store in any case where validation fails.  I still think we'd need to look at the impl to make sure store is the source we use to give results to get queires

Reply:
I'm using this: https://docs.rs/libp2p/latest/libp2p/kad/enum.StoreInserts.html#variant.FilterBoth

Relevant code is here: https://github.com/libp2p/rust-libp2p/blob/4725f46f0d0e14f3ef4e88edc085729c1312dff6/protocols/kad/src/behaviour.rs#L1853-L1893

It basically allows for `PutRequest` middleware. I can make a test for rejection

Reply:
Or maybe we can use: https://docs.rs/libp2p/latest/libp2p/kad/enum.StoreInserts.html

Reply:
Oh lol I just linked it, I missed this part in our code, if you are using that code then why do we even need the store code here at all?

Reply:
Right, it receives the request here, `FilterBoth` moves the onus for storage to the receiver of the event. So basically a rejection is "not adding it" after receiving the message at this step

Reply:
Sorry I should have read the full doc first, I get it now, I thought we'd have to pass a trait object or something into the libp2p impl but it's just a flag that stores nothing by default.  I guess the only thing to confirm is that the `RecordStore` is the source of truth for get queries (I think it is IIRC)

Reply:
Yeah, the local store is `RecordStore`: https://github.com/EspressoSystems/HotShot/blob/b2f9ee7950087e0ce9d9c0fd2ca0a16deef10637/crates/libp2p-networking/src/network/behaviours/dht/mod.rs#L184

Reply:
This is our code, but I mean if it's not already in our store, does node receiving our get query retrieve the record from the store, or some other place?

Reply:
Looks like it uses the store so this is all good: https://docs.rs/libp2p-kad/0.45.3/src/libp2p_kad/behaviour.rs.html#2293